clang/lib/Parse/ParseStmt.cpp

```diff
@@ -496,9 +496,31 @@ StmtResult Parser::ParseExprStatement(ParsedStmtContext StmtCtx) {
     return ParseCaseStatement(StmtCtx, /*MissingCase=*/true, Expr);
   }
 
+  // Parse where clause on an ExprStmt, if it exists.
+  WhereClause *WClause = nullptr;
+  bool WhereClauseExists = false;
+  if (StartsWhereClause(Tok)) {
+    WClause = ParseWhereClause();
+    if (!WClause)
+      return StmtError();
+    WhereClauseExists = true;
+  }
+
   // Otherwise, eat the semicolon.
   ExpectAndConsumeSemi(diag::err_expected_semi_after_expr);
-  return handleExprStmt(Expr, StmtCtx);
+
+  StmtResult StmtRes = handleExprStmt(Expr, StmtCtx);
+  if (StmtRes.isInvalid() || !WhereClauseExists)
+    return StmtRes;
+
+  ValueStmt *VS = dyn_cast<ValueStmt>(StmtRes.get());
+  if (!VS) {
+    Diag(Tok, diag::err_invalid_stmt_where_clause);
```

Comment:
I think it's better to emit llvm_unreachable here instead of emitting a diagnostic.